File: backend/app/schemas/utility_alarm.py
"""
Utility Alarm & Anomaly Detection — Pydantic v2 Schemas
"""
from __future__ import annotations

from datetime import datetime
from typing import Dict, List, Optional
from uuid import UUID

from pydantic import BaseModel, ConfigDict


# ── Alarm Rule schemas ─────────────────────────────────────────────────────────

class AlarmRuleBase(BaseModel):
    rule_code: str
    name: str
    description: Optional[str] = None
    utility_type: str
    device_id: Optional[UUID] = None
    asset_id: Optional[UUID] = None
    alarm_type: Optional[str] = None
    source_module: Optional[str] = None
    alarm_category: str = "GENERAL"
    detection_type: str = "THRESHOLD"
    parameter: str
    operator: str
    threshold_value: float
    threshold_unit: Optional[str] = None
    rolling_window_hours: Optional[int] = None
    delta_pct_threshold: Optional[float] = None
    standard_value: Optional[float] = None
    comparison_asset_id: Optional[UUID] = None
    severity: str = "WARNING"
    cooldown_minutes: int = 15
    evaluation_frequency_min: int = 60
    is_active: bool = True
    notification_emails: Optional[str] = None
    notification_channel: Optional[str] = None
    notes: Optional[str] = None


class AlarmRuleCreate(AlarmRuleBase):
    pass


class AlarmRuleUpdate(BaseModel):
    rule_code: Optional[str] = None
    name: Optional[str] = None
    description: Optional[str] = None
    utility_type: Optional[str] = None
    device_id: Optional[UUID] = None
    asset_id: Optional[UUID] = None
    alarm_type: Optional[str] = None
    source_module: Optional[str] = None
    alarm_category: Optional[str] = None
    detection_type: Optional[str] = None
    parameter: Optional[str] = None
    operator: Optional[str] = None
    threshold_value: Optional[float] = None
    threshold_unit: Optional[str] = None
    rolling_window_hours: Optional[int] = None
    delta_pct_threshold: Optional[float] = None
    standard_value: Optional[float] = None
    comparison_asset_id: Optional[UUID] = None
    severity: Optional[str] = None
    cooldown_minutes: Optional[int] = None
    evaluation_frequency_min: Optional[int] = None
    is_active: Optional[bool] = None
    notification_emails: Optional[str] = None
    notification_channel: Optional[str] = None
    notes: Optional[str] = None


class AlarmRuleRead(AlarmRuleBase):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    created_at: datetime
    updated_at: Optional[datetime] = None
    created_by_name: Optional[str] = None
    device_name: Optional[str] = None
    asset_name: Optional[str] = None
    event_count: int = 0


# ── Alarm Event schemas ────────────────────────────────────────────────────────

class AlarmEventBase(BaseModel):
    rule_id: UUID
    device_id: Optional[UUID] = None
    asset_id: Optional[UUID] = None
    triggered_at: datetime
    status: str
    severity: str
    alarm_type: Optional[str] = None
    source_module: Optional[str] = None
    detection_type: Optional[str] = None
    parameter: Optional[str] = None
    triggered_value: Optional[float] = None
    expected_value: Optional[float] = None
    threshold_value: Optional[float] = None
    threshold_unit: Optional[str] = None
    deviation_pct: Optional[float] = None
    notes: Optional[str] = None


class AlarmEventRead(AlarmEventBase):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    event_no: str
    acknowledged_at: Optional[datetime] = None
    resolved_at: Optional[datetime] = None
    root_cause: Optional[str] = None
    corrective_action: Optional[str] = None
    created_at: datetime
    rule_code: Optional[str] = None
    rule_name: Optional[str] = None
    device_name: Optional[str] = None
    asset_name: Optional[str] = None
    assigned_to_name: Optional[str] = None
    acknowledged_by_name: Optional[str] = None
    resolved_by_name: Optional[str] = None


# ── Action schemas ─────────────────────────────────────────────────────────────

class AlarmEventAcknowledge(BaseModel):
    notes: Optional[str] = None


class AlarmEventAssign(BaseModel):
    assigned_to_id: UUID


class AlarmEventResolve(BaseModel):
    root_cause: str
    corrective_action: str
    notes: Optional[str] = None


# ── Summary schema ─────────────────────────────────────────────────────────────

class AlarmSummary(BaseModel):
    total_active: int
    total_acknowledged: int
    critical_count: int
    alert_count: int
    warning_count: int
    info_count: int
    by_utility_type: Dict[str, int]
    by_category: Dict[str, int]


# ── Template schema ────────────────────────────────────────────────────────────

class AlarmRuleTemplate(BaseModel):
    template_key: str
    name: str
    description: str
    alarm_type: str
    source_module: str
    utility_type: str
    alarm_category: str
    detection_type: str
    parameter: str
    operator: str
    threshold_value: float
    threshold_unit: str
    severity: str
    rolling_window_hours: Optional[int] = None
    delta_pct_threshold: Optional[float] = None
    standard_value: Optional[float] = None

File: frontend/src/lib/utilityAlarm.ts
/**
 * Utility Alarm & Anomaly Detection — TypeScript types and API client.
 */
import { apiClient } from "./api";

// ── Enums ─────────────────────────────────────────────────────────────────────

export type AlarmSeverity = "INFO" | "WARNING" | "ALERT" | "CRITICAL";
export type AlarmStatus = "ACTIVE" | "ACKNOWLEDGED" | "RESOLVED" | "SUPPRESSED";
export type AlarmDetectionType =
  | "THRESHOLD"
  | "SUDDEN_JUMP"
  | "ROLLING_DEVIATION"
  | "SHIFT_COMPARISON"
  | "MACHINE_COMPARISON"
  | "VS_STANDARD";
export type AlarmCategory =
  | "ENERGY"
  | "WATER"
  | "BOILER_STEAM"
  | "COMPRESSED_AIR"
  | "SOLAR"
  | "WASTEWATER"
  | "CHEMICAL"
  | "COMPLIANCE"
  | "COST"
  | "GENERAL";
export type AlarmOperator = "GT" | "LT" | "GTE" | "LTE" | "EQ" | "NEQ" | "DELTA_PCT";
export type UtilityType =
  | "ELECTRICITY"
  | "WATER"
  | "SOFT_WATER"
  | "PROCESS_WATER"
  | "STEAM"
  | "COMPRESSED_AIR"
  | "SOLAR"
  | "NATURAL_GAS"
  | "WASTEWATER"
  | "DIESEL"
  | "CHILLED_WATER"
  | "OTHER";

// ── Interfaces ────────────────────────────────────────────────────────────────

export interface AlarmRule {
  id: string;
  rule_code: string;
  name: string;
  description?: string | null;
  utility_type: UtilityType;
  alarm_type?: string | null;
  source_module?: string | null;
  alarm_category: AlarmCategory;
  detection_type: AlarmDetectionType;
  parameter: string;
  operator: AlarmOperator;
  threshold_value: number;
  threshold_unit?: string | null;
  rolling_window_hours?: number | null;
  delta_pct_threshold?: number | null;
  standard_value?: number | null;
  severity: AlarmSeverity;
  cooldown_minutes: number;
  evaluation_frequency_min: number;
  is_active: boolean;
  notification_emails?: string | null;
  notification_channel?: string | null;
  notes?: string | null;
  device_name?: string | null;
  asset_name?: string | null;
  created_by_name?: string | null;
  event_count: number;
  created_at: string;
}

export interface AlarmEvent {
  id: string;
  event_no: string;
  rule_id: string;
  rule_code?: string | null;
  rule_name?: string | null;
  triggered_at: string;
  acknowledged_at?: string | null;
  resolved_at?: string | null;
  status: AlarmStatus;
  severity: AlarmSeverity;
  alarm_type?: string | null;
  source_module?: string | null;
  detection_type?: string | null;
  parameter?: string | null;
  triggered_value?: number | null;
  expected_value?: number | null;
  threshold_value?: number | null;
  threshold_unit?: string | null;
  deviation_pct?: number | null;
  root_cause?: string | null;
  corrective_action?: string | null;
  device_name?: string | null;
  asset_name?: string | null;
  assigned_to_name?: string | null;
  acknowledged_by_name?: string | null;
  resolved_by_name?: string | null;
  notes?: string | null;
  created_at: string;
}

export interface AlarmSummary {
  total_active: number;
  total_acknowledged: number;
  critical_count: number;
  alert_count: number;
  warning_count: number;
  info_count: number;
  by_utility_type: Record<string, number>;
  by_category: Record<string, number>;
}

export interface AlarmRuleTemplate {
  template_key: string;
  name: string;
  description: string;
  alarm_type: string;
  source_module: string;
  utility_type: UtilityType;
  alarm_category: AlarmCategory;
  detection_type: AlarmDetectionType;
  parameter: string;
  operator: AlarmOperator;
  threshold_value: number;
  threshold_unit: string;
  severity: AlarmSeverity;
  rolling_window_hours?: number | null;
  delta_pct_threshold?: number | null;
  standard_value?: number | null;
}

export interface AlarmRuleFilters {
  utility_type?: UtilityType;
  severity?: AlarmSeverity;
  is_active?: boolean;
  detection_type?: AlarmDetectionType;
  skip?: number;
  limit?: number;
}

export interface AlarmEventFilters {
  status?: AlarmStatus;
  severity?: AlarmSeverity;
  utility_type?: UtilityType;
  date_from?: string;
  date_to?: string;
  skip?: number;
  limit?: number;
}

export type AlarmRuleCreate = Omit<AlarmRule, "id" | "event_count" | "created_at" | "created_by_name">;
export type AlarmRuleUpdate = Partial<Omit<AlarmRule, "id" | "rule_code" | "event_count" | "created_at" | "created_by_name">>;

// ── Color maps ────────────────────────────────────────────────────────────────

export const SEVERITY_COLOR: Record<AlarmSeverity, string> = {
  CRITICAL: "#ef4444",
  ALERT:    "#f97316",
  WARNING:  "#eab308",
  INFO:     "#60a5fa",
};

export const SEVERITY_BG_CLASS: Record<AlarmSeverity, string> = {
  CRITICAL: "bg-red-600 text-white",
  ALERT:    "bg-orange-500 text-white",
  WARNING:  "bg-yellow-500 text-white",
  INFO:     "bg-blue-500 text-white",
};

export const SEVERITY_SOFT_CLASS: Record<AlarmSeverity, string> = {
  CRITICAL: "bg-red-100 text-red-700 dark:bg-red-900/30 dark:text-red-400",
  ALERT:    "bg-orange-100 text-orange-700 dark:bg-orange-900/30 dark:text-orange-400",
  WARNING:  "bg-yellow-100 text-yellow-700 dark:bg-yellow-900/30 dark:text-yellow-400",
  INFO:     "bg-blue-100 text-blue-700 dark:bg-blue-900/30 dark:text-blue-400",
};

export const STATUS_SOFT_CLASS: Record<AlarmStatus, string> = {
  ACTIVE:       "bg-red-100 text-red-700 dark:bg-red-900/30 dark:text-red-400",
  ACKNOWLEDGED: "bg-yellow-100 text-yellow-700 dark:bg-yellow-900/30 dark:text-yellow-400",
  RESOLVED:     "bg-green-100 text-green-700 dark:bg-green-900/30 dark:text-green-400",
  SUPPRESSED:   "bg-gray-100 text-gray-500 dark:bg-gray-700 dark:text-gray-400",
};

// ── Label maps ────────────────────────────────────────────────────────────────

export const DETECTION_TYPE_LABELS: Record<AlarmDetectionType, string> = {
  THRESHOLD:          "Threshold",
  SUDDEN_JUMP:        "Sudden Jump",
  ROLLING_DEVIATION:  "Rolling Avg",
  SHIFT_COMPARISON:   "Shift Comparison",
  MACHINE_COMPARISON: "Machine Compare",
  VS_STANDARD:        "vs. Standard",
};

export const DETECTION_TYPE_DESCRIPTIONS: Record<AlarmDetectionType, string> = {
  THRESHOLD:          "Triggers when value exceeds a fixed threshold",
  SUDDEN_JUMP:        "Triggers on rapid percentage change between readings",
  ROLLING_DEVIATION:  "Triggers when value deviates from rolling average",
  SHIFT_COMPARISON:   "Compares current shift data to previous shift",
  MACHINE_COMPARISON: "Compares machine performance against similar machines",
  VS_STANDARD:        "Triggers when value deviates from standard/benchmark",
};

export const ALARM_CATEGORY_LABELS: Record<AlarmCategory, string> = {
  ENERGY:        "Energy",
  WATER:         "Water",
  BOILER_STEAM:  "Boiler / Steam",
  COMPRESSED_AIR:"Compressed Air",
  SOLAR:         "Solar",
  WASTEWATER:    "Wastewater",
  CHEMICAL:      "Chemical",
  COMPLIANCE:    "Compliance",
  COST:          "Cost",
  GENERAL:       "General",
};

export const OPERATOR_LABELS: Record<AlarmOperator, string> = {
  GT:        "Greater than (>)",
  LT:        "Less than (<)",
  GTE:       "Greater than or equal (≥)",
  LTE:       "Less than or equal (≤)",
  EQ:        "Equals (=)",
  NEQ:       "Not equal (≠)",
  DELTA_PCT: "Delta percentage (%Δ)",
};

export const OPERATOR_SYMBOLS: Record<AlarmOperator, string> = {
  GT:        ">",
  LT:        "<",
  GTE:       "≥",
  LTE:       "≤",
  EQ:        "=",
  NEQ:       "≠",
  DELTA_PCT: "%Δ",
};

export const UTILITY_TYPE_LABELS: Record<UtilityType, string> = {
  ELECTRICITY:    "Electricity",
  WATER:          "Water",
  SOFT_WATER:     "Soft Water",
  PROCESS_WATER:  "Process Water",
  STEAM:          "Steam",
  COMPRESSED_AIR: "Compressed Air",
  SOLAR:          "Solar",
  NATURAL_GAS:    "Natural Gas",
  WASTEWATER:     "Wastewater",
  DIESEL:         "Diesel",
  CHILLED_WATER:  "Chilled Water",
  OTHER:          "Other",
};

export const UTILITY_TYPE_ICONS: Record<UtilityType, string> = {
  ELECTRICITY:    "⚡",
  WATER:          "💧",
  SOFT_WATER:     "🌊",
  PROCESS_WATER:  "🔵",
  STEAM:          "♨️",
  COMPRESSED_AIR: "💨",
  SOLAR:          "☀️",
  NATURAL_GAS:    "🔥",
  WASTEWATER:     "♻️",
  DIESEL:         "🛢️",
  CHILLED_WATER:  "❄️",
  OTHER:          "🔧",
};

export const ALARM_CATEGORY_ICONS: Record<AlarmCategory, string> = {
  ENERGY:         "⚡",
  WATER:          "💧",
  BOILER_STEAM:   "♨️",
  COMPRESSED_AIR: "💨",
  SOLAR:          "☀️",
  WASTEWATER:     "♻️",
  CHEMICAL:       "🧪",
  COMPLIANCE:     "📋",
  COST:           "💰",
  GENERAL:        "🔔",
};

export const SEVERITY_ICONS: Record<AlarmSeverity, string> = {
  CRITICAL: "🔴",
  ALERT:    "🟠",
  WARNING:  "⚠",
  INFO:     "ℹ",
};

// ── API client ────────────────────────────────────────────────────────────────

export const alarmApi = {
  // Rules
  async listRules(params?: AlarmRuleFilters): Promise<AlarmRule[]> {
    const filtered = params
      ? Object.fromEntries(Object.entries(params).filter(([, v]) => v !== undefined && v !== ""))
      : {};
    const r = await apiClient.get<AlarmRule[]>("/api/v1/alarms/rules", { params: filtered }).then((r) => r.data);
    return r.data;
  },

  async getRuleTemplates(): Promise<AlarmRuleTemplate[]> {
    const r = await apiClient.get<AlarmRuleTemplate[]>("/api/v1/alarms/rules/templates").then((r) => r.data);
    return r.data;
  },

  async createRule(data: AlarmRuleCreate): Promise<AlarmRule> {
    const r = await apiClient.post<AlarmRule>("/api/v1/alarms/rules", data).then((r) => r.data);
    return r.data;
  },

  async getRule(id: string): Promise<AlarmRule> {
    const r = await apiClient.get<AlarmRule>(`/api/v1/alarms/rules/${id}`).then((r) => r.data);
    return r.data;
  },

  async updateRule(id: string, data: AlarmRuleUpdate): Promise<AlarmRule> {
    const r = await apiClient.patch<AlarmRule>(`/api/v1/alarms/rules/${id}`, data).then((r) => r.data);
    return r.data;
  },

  async deleteRule(id: string): Promise<void> {
    await apiClient.delete(`/api/v1/alarms/rules/${id}`).then((r) => r.data);
  },

  async toggleRule(id: string): Promise<AlarmRule> {
    const r = await apiClient.post<AlarmRule>(`/api/v1/alarms/rules/${id}/toggle`).then((r) => r.data);
    return r.data;
  },

  // Events
  async getEventsSummary(): Promise<AlarmSummary> {
    const r = await apiClient.get<AlarmSummary>("/api/v1/alarms/events/summary").then((r) => r.data);
    return r.data;
  },

  async listEvents(params?: AlarmEventFilters): Promise<AlarmEvent[]> {
    const filtered = params
      ? Object.fromEntries(Object.entries(params).filter(([, v]) => v !== undefined && v !== ""))
      : {};
    const r = await apiClient.get<AlarmEvent[]>("/api/v1/alarms/events", { params: filtered }).then((r) => r.data);
    return r.data;
  },

  async getEvent(id: string): Promise<AlarmEvent> {
    const r = await apiClient.get<AlarmEvent>(`/api/v1/alarms/events/${id}`).then((r) => r.data);
    return r.data;
  },

  async acknowledgeEvent(id: string, notes?: string): Promise<AlarmEvent> {
    const r = await apiClient.post<AlarmEvent>(`/api/v1/alarms/events/${id}/acknowledge`, { notes }).then((r) => r.data);
    return r.data;
  },

  async assignEvent(id: string, assigned_to_id: string): Promise<AlarmEvent> {
    const r = await apiClient.post<AlarmEvent>(`/api/v1/alarms/events/${id}/assign`, { assigned_to_id }).then((r) => r.data);
    return r.data;
  },

  async resolveEvent(
    id: string,
    payload: { root_cause: string; corrective_action: string; notes?: string },
  ): Promise<AlarmEvent> {
    const r = await apiClient.post<AlarmEvent>(`/api/v1/alarms/events/${id}/resolve`, payload).then((r) => r.data);
    return r.data;
  },

  async suppressEvent(id: string): Promise<AlarmEvent> {
    const r = await apiClient.post<AlarmEvent>(`/api/v1/alarms/events/${id}/suppress`).then((r) => r.data);
    return r.data;
  },

  // Evaluation
  async evaluate(): Promise<{ new_alarms: number; message: string }> {
    const r = await apiClient.post<{ new_alarms: number; message: string }>("/api/v1/alarms/evaluate").then((r) => r.data);
    return r.data;
  },
};
